enclave/src/polkadex_gateway.rs

```diff
@@ -0,0 +1,293 @@
+                debug!("Main Account is not registered");
+                return Err(GatewayError::MainAccountNotRegistered);
+            }
+        }
+    }
+    Ok(())
+}
+
+static CREATE_ORDER_NONCE: AtomicPtr<()> = AtomicPtr::new(0 as *mut ());
+static CREATE_ORDER_CACHE: AtomicPtr<()> = AtomicPtr::new(0 as *mut ());
+static CANCEL_ORDER_CACHE: AtomicPtr<()> = AtomicPtr::new(0 as *mut ());
+
+pub fn initialize_polkadex_gateway() {
+    let nonce: u128 = 0;
+    let create_nonce_storage_ptr = Arc::new(SgxMutex::<u128>::new(nonce));
+    let create_nonce_ptr = Arc::into_raw(create_nonce_storage_ptr);
+    CREATE_ORDER_NONCE.store(create_nonce_ptr as *mut (), Ordering::SeqCst);
+
+    let cancel_cache: HashSet<OrderUUID> = HashSet::new();
+    let cancel_cache_storage_ptr = Arc::new(SgxMutex::new(cancel_cache));
+    let cancel_cache_ptr = Arc::into_raw(cancel_cache_storage_ptr);
+    CANCEL_ORDER_CACHE.store(cancel_cache_ptr as *mut (), Ordering::SeqCst);
+
+    let create_cache: HashMap<u128, Order> = HashMap::new();
+    let create_cache_storage_ptr = Arc::new(SgxMutex::new(create_cache));
+    let create_cache_ptr = Arc::into_raw(create_cache_storage_ptr);
+    CREATE_ORDER_CACHE.store(create_cache_ptr as *mut (), Ordering::SeqCst);
+}
+
+fn load_nonce_pointer() -> SgxResult<&'static SgxMutex<u128>> {
+    let ptr = CREATE_ORDER_NONCE.load(Ordering::SeqCst) as *mut SgxMutex<u128>;
+    if ptr.is_null() {
+        error!("Pointer is Null");
+        return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+    } else {
+        Ok(unsafe { &*ptr })
+    }
+}
+
+fn load_cancel_cache_pointer() -> SgxResult<&'static SgxMutex<HashSet<OrderUUID>>> {
+    let ptr = CANCEL_ORDER_CACHE.load(Ordering::SeqCst) as *mut SgxMutex<HashSet<OrderUUID>>;
+    if ptr.is_null() {
+        error!("Pointer is Null");
+        return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+    } else {
+        Ok(unsafe { &*ptr })
+    }
+}
+
+fn load_create_cache_pointer() -> SgxResult<&'static SgxMutex<HashMap<u128, Order>>> {
+    let ptr = CREATE_ORDER_CACHE.load(Ordering::SeqCst) as *mut SgxMutex<HashMap<u128, Order>>;
+    if ptr.is_null() {
+        error!("Pointer is Null");
+        return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+    } else {
+        Ok(unsafe { &*ptr })
+    }
+}
+
+fn get_new_nonce() -> SgxResult<u128> {
+    let mutex = load_nonce_pointer()?;
+    let mut nonce: SgxMutexGuard<u128> = mutex.lock().unwrap();
+    let current_nonce = nonce.clone();
+    nonce.saturating_add(1);
+    Ok(current_nonce)
+}
+
+
+pub unsafe fn remove_order_from_cache_and_store_in_ordermirror(nonce: u128, order_uuid: OrderUUID) -> SgxResult<()> {
+    let mutex = load_create_cache_pointer()?;
+    let cache: SgxMutexGuard<HashMap<u128, Order>> = mutex.lock().unwrap();
+    if let Some(order) = cache.get(&nonce) {
+        polkadex_orderbook_storage::add_order(order.clone(), order_uuid)?;
```

Comment:
I think the `remove_order_from_cache` part is missing. Maybe add a TODO?

Reply:
I have added a TODO since we are moving ahead with threading model and not caching model of handling RPC calls